### PR TITLE
feat(portal): add initial portal foundation

### PR DESCRIPTION
**Pull request by jelbourn**
R: @hansl @kara 

See the [design doc](https://docs.google.com/document/d/1jSMjkKSdzbQa2cuCeSFpzFAwDHr7xY4eqsXkfF6MeqA/preview)

### CODE REVIEW

File: src/core/portal/portal-directives.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
+import {Portal} from './portal';
```
Comment (hansl) [inline src/core/portal/portal-directives.ts]: Why not put this in `src/directives/portal`? Is this not something we want to expose publicly?

File: src/core/portal/portal-directives.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+ * Usage:
+ * <template [portalHost]="greeting"></template>
+ */
+@Directive({
+  selector: '[portalHost]',
+  inputs: ['portal: portalHost']
+})
+export class PortalHostDirective extends BasePortalHost {
+  /** The attached portal. */
+  private portal_: Portal<any>;
```
Comment (hansl) [inline src/core/portal/portal-directives.ts]: `any` might be a bit too generic? `Portal<Type>` might be better and convey your meaning.

File: src/core/portal/portal-directives.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+  constructor(
+      private dynamicComponentLoader: DynamicComponentLoader,
```
Comment (hansl) [inline src/core/portal/portal-directives.ts]: `dynamicComponentLoader_`?

File: src/core/portal/portal-directives.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  /** Attach the given ComponentPortal to this PortlHost using the DynamicComponentLoader. */
+  attachComponentPortal(portal: ComponentPortal): Promise<ComponentRef> {
+    portal.setAttachedHost(this);
+
+    // If the portal specifies an origin, use that as the logical location of the component
+    // in the application tree. Otherwise use the location of this PortalHost.
+    let elementRef = portal.origin != null ? portal.origin : this.elementRef_;
+
+    // Typecast is necessary for Dart transpilation.
+    return <Promise<ComponentRef>>
```
Comment (hansl) [inline src/core/portal/portal-directives.ts]: It seems like this cast is unnecessary, as `loadNextToLocation` already returns this.

File: src/core/portal/portal-directives.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    let viewRef = this.viewContainerRef_.createEmbeddedView(portal.templateRef);
+    portal.locals.forEach((k, v) => viewRef.setLocal(k, v));
+    this.setDisposeFn(() => this.viewContainerRef_.clear());
+
+    // TODO(jelbourn): return locals from view
+    // Typecast is necessary for Dart transpilation.
+    return <Promise<Map<string, any>>> PromiseWrapper.resolve(new Map<string, any>());
+  }
+
+  /** Detatches the currently attached Portal (if there is one) and attaches the given Portal. */
```
Comment (hansl) [inline src/core/portal/portal-directives.ts]: typo: `detaches`.

File: src/core/portal/portal-directives.ts (hunk 1 above)
Comment (jelbourn) [inline src/core/portal/portal-directives.ts]: The general best practice is to organize code by feature / area than by type of thing it is (directive, service, etc.). As such, all the portal code should go together. This would be part of the general component-building toolkit.

File: src/core/portal/portal-directives.ts (hunk 2 above)
Comment (jelbourn) [inline src/core/portal/portal-directives.ts]: For a `TemplatePortal` the generic is `Map<string, any>` rather than `Type`

File: src/core/portal/portal-directives.ts (hunk 3 above)
Comment (jelbourn) [inline src/core/portal/portal-directives.ts]: Done

File: src/core/portal/portal-directives.ts (hunk 4 above)
Comment (jelbourn) [inline src/core/portal/portal-directives.ts]: And yet DartAnalyzer fails completely without it.

File: src/core/portal/portal-directives.ts (hunk 5 above)
Comment (jelbourn) [inline src/core/portal/portal-directives.ts]: Done.

**Comment by hansl** [comment]
README.md could be really nice, even if it's simple.

**Comment by jelbourn** [comment]
@hansl Added a readme. 

cc @pkozlowski-opensource since you've been looking at the design doc.

File: src/core/portal/README.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+<p *portal>
+  The content of this template is captured by the portal.
+</p>
+```
+
+A component can use `@ViewChild` or `@ViewChildren` to get a reference to a 
+`TemplatePortalDiective`.
+
+##### `ComponentPortal`
+Used to create a portal form a component type.
```
Comment (kara) [inline src/core/portal/README.md]: Typo?  `form` -> `from`

File: src/core/portal/portal-directives.ts [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  private replaceAttachedPortal_(p: Portal<any>): void {
+    if (this.hasAttached()) {
+      this.detach().then(_ => {
+        if (p != null) {
+          this.attach(p);
+          this.portal_ = p;
+        }
+      });
```
Comment (kara) [inline src/core/portal/portal-directives.ts]: Nitpicky, but these four lines (98-102) are repeated in 104-107. Perhaps they should be abstracted into a little method?

File: src/core/portal/README.md (hunk 6 above)
Comment (jelbourn) [inline src/core/portal/README.md]: Done.

File: src/core/portal/portal.ts [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+      this.attachedPortal_ = portal;
+      return this.attachComponentPortal(portal);
+    } else if (portal instanceof TemplatePortal) {
+      this.attachedPortal_ = portal;
+      return this.attachTemplatePortal(portal);
+    }
+
+    throw new BaseException(
+        'Attempting to attach an unknown Portal type. ' +
+        'BasePortalHost accepts either a ComponentPortal or a TemplatePortal.');
```
Comment (kara) [inline src/core/portal/portal.ts]: Maybe we could use a template string here for multi-line rather than concatenation?

**Comment by kara** [comment]
Made some nitpicky comments, but LGTM

File: src/core/portal/portal.ts (hunk 8 above)
Comment (jelbourn) [inline src/core/portal/portal.ts]: I originally had a template string, but the line breaks carried into the error message, which made it not show the whole message without expansion in the dev tools.

File: src/core/portal/portal-directives.ts (hunk 7 above)
Comment (jelbourn) [inline src/core/portal/portal-directives.ts]: Excellent point; I simplified it.

File: src/core/portal/portal.ts (hunk 8 above)
Comment (kara) [inline src/core/portal/portal.ts]: Ah, that makes sense then.

**Comment by jelbourn** [comment]
Fixed a new emo bug that popped up (angular2-polyfills was loaded in the wrong place) and added another test for _changing_ the portal.

File: src/core/portal/portal-directives.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+import {TemplatePortal} from './portal';
+import {ComponentRef} from 'angular2/core';
+import {ComponentPortal} from './portal';
+import {Directive} from 'angular2/core';
+import {TemplateRef} from 'angular2/core';
+import {BasePortalHost} from './portal';
+import {DynamicComponentLoader} from 'angular2/core';
+import {ElementRef} from 'angular2/core';
+import {ViewContainerRef} from 'angular2/core';
+import {PromiseWrapper} from 'angular2/src/facade/promise';
```
Comment (hansl) [inline src/core/portal/portal-directives.ts]: You shouldn't need this.

File: src/core/portal/portal.ts [hunk 10]
```diff
@@ -0,0 +1,194 @@
+import {TemplateRef, Type} from 'angular2/core';
+import {ElementRef} from 'angular2/core';
+import {ComponentRef} from 'angular2/core';
+
+import {BaseException} from 'angular2/src/facade/exceptions';
+import {PromiseWrapper} from 'angular2/src/facade/promise';
```
Comment (hansl) [inline src/core/portal/portal.ts]: You should not need this.

File: src/core/portal/portal.ts [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+    super();
+    this.templateRef = template;
+  }
+
+  get origin(): ElementRef {
+    return this.templateRef.elementRef;
+  }
+
+  attach(host: PortalHost, locals?: Map<string, any>): Promise<Map<string, any>> {
+    this.locals = locals == null ? new Map<string, any>() : locals;
```
Comment (hansl) [inline src/core/portal/portal.ts]: `isPresent(locals)`?

File: src/core/portal/portal.ts [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+
+
+/**
+ * A `TemplatePortal` is a portal that represents some embedded template (TemplateRef).
+ */
+export class TemplatePortal extends Portal<Map<string, any>> {
+  /** The embedded template that will be used to instantiate an embedded View in the host. */
+  templateRef: TemplateRef;
+
+  /** Additional locals for the instantiated embedded view. */
```
Comment (hansl) [inline src/core/portal/portal.ts]: What are `locals`? Just scoped values to share between the portal and its host?

File: src/core/portal/portal.spec.ts [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+
+      testAppComponent.selectedPortal = new ComponentPortal(PizzaMsg);
+      appFixture.detectChanges();
+
+      flushMicrotasks();
+
+      expect(hostContainer.textContent).toContain('Pizza');
+    }));
+
+  });
```
Comment (hansl) [inline src/core/portal/portal.spec.ts]: Regression of testing for the behaviour expected of `locals`?

File: src/core/portal/README.md [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+| `detach(): Promise<void>` | Detaches the portal from its host. |
+| `isAttached: boolean` | Whether the portal is attached. |
+
+##### `PortalHost`
+| Method | Description |
+| --- | --- |
+| `attach(Portal): Promise<void>` | Attaches a portal to the host. |
+| `detach(): Promise<void>` | Detaches the portal from the host. |
+| `dispose(): Promise<void>` | Permanently dispose the host. |
+| `hasattached: boolean` | Whether a portal is attached to the host. |
```
Comment (hansl) [inline src/core/portal/README.md]: `hasAttached()`

File: src/core/portal/portal.ts (hunk 12 above)
Comment (jelbourn) [inline src/core/portal/portal.ts]: They're extra locals passed to the [`EmbeddedView`](https://angular.io/docs/ts/latest/api/core/EmbeddedViewRef-class.html) via `setLocals`. I added a unit test where you can see how they're used.

File: src/core/portal/portal.ts (hunk 11 above)
Comment (jelbourn) [inline src/core/portal/portal.ts]: Why change it? Using `== null` is idomatic TS / JS that has the wonderfully rare benefit of also working in Dart.

File: src/core/portal/portal-directives.ts (hunk 9 above)
Comment (jelbourn) [inline src/core/portal/portal-directives.ts]: Done.

File: src/core/portal/portal.ts (hunk 10 above)
Comment (jelbourn) [inline src/core/portal/portal.ts]: Done.

File: src/core/portal/portal.spec.ts (hunk 13 above)
Comment (jelbourn) [inline src/core/portal/portal.spec.ts]: Well, this sent me down a long, long journey of trying to get the test working for this. There _was_ a bug, which was that I had `k` and `v` swapped in my `forEach`.

However, there is triumph, and there's now  a unit test for the `locals` behavior as well as a test for a `TemplatePortal` that contains a binding.

File: src/core/portal/README.md (hunk 14 above)
Comment (jelbourn) [inline src/core/portal/README.md]: Done.

**Comment by jelbourn** [comment]
@hansl all comments should be addressed now. Good call on the `locals` thing, it had slipped my mind.

**Comment by hansl** [comment]
Could you add a link to https://angular.io/docs/ts/latest/api/core/EmbeddedViewRef-class.html for the locals? I still find "local" ridiculously unfamiliar. Finally, if you can change `Portal<any>` to maybe `Portal<ComponentRef>` (or something close). `Portal<number>` doesn't make sense ;)

Otherwise, LGTM :+1:

**Comment by jelbourn** [comment]
Done on the description for `locals`. As discussed, the `any` on `Portal` make sense because it really can be anything (e.g., `Portal<number>` would be very simple, but still make sense).
